### PR TITLE
feat: add idp templates

### PR DESCRIPTION
**Pull request by eliobischof**
- [x] Adds Instance IDP Templates
- [x] Adds Org IDP Templates
- [x] Enables testing them with `TF_ACC=1 TF_ACC_ZITADEL_TOKEN=<my_json_token> go test -v ./...`
- [x] Makes them addable to login policies
- [x] Adds other templates
- [x] https://github.com/zitadel/terraform-provider-zitadel/pull/82
- [x] Updates dependencies
- [x] Cleanup
- [x] Adds CONTRIBUTING.md

### CODE REVIEW

**Comment by eliobischof** [comment]
@adlerhurst I've implemented the GitHub org and instance templates. Will roll them out now.
The PR is ready for an early review.

File: zitadel/v2/idp_azure_ad/funcs.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+		tenant.Type = &idp.AzureADTenant_TenantId{
+			TenantId: tenantId,
+		}
+	} else {
+		tenant.Type = &idp.AzureADTenant_TenantType{
+			TenantType: idp.AzureADTenantType(idp.AzureADTenantType_value[d.Get(idp_utils.TenantTypeVar).(string)]),
+		}
+	}
+	return tenant
+}
```
Comment (eliobischof) [inline zitadel/v2/idp_azure_ad/funcs.go:133]: Is the implementation logic correct?
If a tenant ID is configured, we use the tentant ID for create and update requests, else we use the tentant type.

File: zitadel/v2/idp_azure_ad/datasource.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/zitadel/terraform-provider-zitadel/zitadel/v2/idp_utils"
+)
+
+func GetDatasource() *schema.Resource {
+	return &schema.Resource{
+		Description: "Datasource representing an Azure AD IDP on the instance.",
+		Schema: map[string]*schema.Schema{
+			idp_utils.IdpIDVar: {
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_azure_ad/datasource.go:12]: You're right, schema variables can be defines centrally, too. Will do that

File: zitadel/v2/idp_azure_ad/resource.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+				Required:    true,
+				Description: "client secret generated by the identity provider",
+				Sensitive:   true,
+			},
+			idp_utils.ScopesVar: {
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Required:    true,
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_azure_ad/resource.go:36]: I guess not, forgot to recheck that. I'll have a look

File: zitadel/v2/idp_github/resource.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	"github.com/zitadel/terraform-provider-zitadel/zitadel/v2/idp_utils"
+)
+
+func GetResource() *schema.Resource {
+	return &schema.Resource{
+		Description: "Resource representing a GitHub IDP on the instance.",
+		Schema: map[string]*schema.Schema{
+			idp_utils.NameVar: {
+				Type:        schema.TypeString,
+				Required:    true,
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_github/resource.go:15]: Will check

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+   
+   # go to a directory containing .tf files.
+   cd /my-zitadel-terraform-files
+   
+   # apply them
+   terraform apply
+   ```
+4. The execution stops at your breakpoints.
+
+# Run Acceptance Tests
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply CONTRIBUTING.md:18]: Will mention it.
There is probably more to do regarding tests anyway, like for example running them in a pipeline.

File: zitadel/v2/idp_azure_ad/resource_test.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package idp_azure_ad_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/zitadel/terraform-provider-zitadel/zitadel/v2/idp_utils/test_utils"
+)
+
+func TestAccZITADELInstanceIdPAzureAD(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_azure_ad/resource_test.go:10]: Testing enums is prpbably worth it.
But required vs optional is just a (partly central, see above) decision in the schema and IMO not worth testing. Do you agree?

File: CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (eliobischof) [reply CONTRIBUTING.md:18]: done

File: zitadel/v2/idp_azure_ad/datasource.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_azure_ad/datasource.go]: I moved common schema definitions to the utils package and specific const declarations to provider packages. Schemas look much cleaner now :+1:

File: zitadel/v2/idp_azure_ad/funcs.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+func update(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	clientinfo, ok := m.(*helper.ClientInfo)
+	if !ok {
+		return diag.Errorf("failed to get client")
+	}
+	client, err := helper.GetAdminClient(clientinfo)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if d.HasChangesExcept(idp_utils.IdpIDVar) {
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_azure_ad/funcs.go]: I also removed the checks for org IDs as they have ForceNew=true

File: zitadel/v2/idp_github/resource.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_github/resource.go]: made it optional

File: zitadel/v2/idp_azure_ad/resource.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_azure_ad/resource.go]: made it optional

File: zitadel/v2/idp_azure_ad/funcs.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+		_, err = client.UpdateAzureADProvider(ctx, &admin.UpdateAzureADProviderRequest{
+			Id:            d.Id(),
+			Name:          d.Get(idp_utils.NameVar).(string),
+			ClientId:      d.Get(idp_utils.ClientIDVar).(string),
+			ClientSecret:  d.Get(idp_utils.ClientSecretVar).(string),
+			Scopes:        helper.GetOkSetToStringSlice(d, idp_utils.ScopesVar),
+			Tenant:        ConstructTenant(d),
+			EmailVerified: d.Get(idp_utils.EmailVerifiedVar).(bool),
+			ProviderOptions: &idp.Options{
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_azure_ad/funcs.go]: done

File: zitadel/v2/idp_azure_ad/resource_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (eliobischof) [reply zitadel/v2/idp_azure_ad/resource_test.go:11]: What do you mean by validation of enum attributes? Shouldn't we leave this responsibility to the API?

[the remark this answers is not in the document]
Comment (stebenz) [reply zitadel/v2/idp_azure_ad/resource_test.go:11]: Testing of required and optional attributes is sometimes dependent on the logic in the provider as sometimes the attributes are defined as optional. Still, a combination of different attributes is required, and so on.

True that it is a decision in the schema but sometimes also the logic changes depending on what attributes are set, so I don't agree fully

There is the option to validate values as defined in the schema, for example, look at the application resources, as you use the map defined in the API, if you use no validated value you will get a problem in the ConstructTenant func.

**Review by stebenz — APPROVED** [review]

**Pull request merged**